lib/iopromise/view_component/data_loader.rb

```diff
@@ -1,56 +1,15 @@
 # frozen_string_literal: true
 
+require_relative "../data_loader"
 require "view_component/engine"
 
 module IOPromise
   module ViewComponent
     module DataLoader
-      module ClassMethods
-        def attr_promised_data(*args)
-          @promised_data ||= []
-          @promised_data.concat(args)
-
-          args.each do |arg|
-            self.class_eval("def #{arg};@#{arg}.sync;end")
-          end
-        end
-    
-        def promised_data_keys
-          @promised_data ||= []
-        end
-      end
+      include ::IOPromise::DataLoader
 
       def self.included(base)
-        base.extend(ClassMethods)
-      end
-
-      def data_as_promise
-        @data_promise ||= begin
-          promises = self.class.promised_data_keys.map do |k|
-            p = instance_variable_get('@' + k.to_s)
-            if p.is_a?(IOPromise::ViewComponent::DataLoader)
-              # recursively preload all nested data
-              p.data_as_promise
-            else
-              # for any local promises, we'll unwrap them before completing
-              p.then do |result|
-                if result.is_a?(IOPromise::ViewComponent::DataLoader)
-                  # likewise, if we resolved a promise that we can recurse, load that data too.
-                  result.data_as_promise
-                else
-                  result
-                end
-              end
-            end
-          end
-
-          Promise.all(promises)
-        end
-      end
-
-      def sync
-        data_as_promise.sync
-        self
+        base.extend(::IOPromise::DataLoader::ClassMethods)
```

Comment:
Is this needed with the `included` callback on the `::IOPromise::DataLoader` class above?

Reply:
It surprised me that it was needed by gut instinct too, but it seems to be - I guess it sort of makes sense because by the time it's finally included only the second module is being included, whereas the first has by this point already been included - so there's no opportunity to extend the final class:
```ruby
module IOPromise
  module DataLoader
    module ClassMethods
      def foo
        puts 'foo'
      end
    end

    def self.included(base)
      base.extend(ClassMethods)
    end
  end
end

module IOPromise
  module ViewComponent
    module DataLoader
      include ::IOPromise::DataLoader

      def self.included(base)
        # base.extend(::IOPromise::DataLoader::ClassMethods)
      end
    end
  end
end

class Loader
  include IOPromise::ViewComponent::DataLoader

  foo
end
```

```
$ ruby test.rb
test.rb:30:in `<class:Loader>': undefined local variable or method `foo' for Loader:Class (NameError)
	from test.rb:27:in `<main>'
```